flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/StateMetadata.java

```diff
@@ -0,0 +1,196 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.nodes.exec;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.configuration.ReadableConfig;
+import org.apache.flink.table.api.config.ExecutionConfigOptions;
+import org.apache.flink.util.CollectionUtil;
+import org.apache.flink.util.Preconditions;
+
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonCreator;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import org.apache.flink.shaded.jackson2.com.fasterxml.jackson.annotation.JsonProperty;
+
+import javax.annotation.Nullable;
+
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+/**
+ * It is used to describe the state metadata of a stateful operator, which is
+ * serialized/deserialized into/from those {@link
+ * org.apache.flink.table.planner.plan.nodes.exec.stream.StreamExecNode}s that can generate stateful
+ * operators. For ExecNodes that generates {@link
+ * org.apache.flink.streaming.api.operators.TwoInputStreamOperator} or {@link
+ * org.apache.flink.streaming.api.operators.MultipleInputStreamOperator}, there will be multiple
+ * metadata describing information about each input's state.
+ *
+ * <p>The metadata describes the following attributes.
+ *
+ * <ul>
+ *   <li>{@code stateIndex}: annotates the state is from the i-th input, index based on zero
+ *   <li>{@code ttl}: annotates the state retention time for the i-th input's state, the time unit
+ *       is ms.
+ *   <li>{@code name}: annotates the state description, such as deduplicate-state, join-left-state.
+ * </ul>
+ */
+@Internal
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class StateMetadata {
+    public static final String FIELD_NAME_STATE_INDEX = "index";
+    public static final String FIELD_NAME_STATE_TTL = "ttl";
+    public static final String FIELD_NAME_STATE_NAME = "name";
+
+    @JsonProperty(value = FIELD_NAME_STATE_INDEX, index = 0)
+    private final int stateIndex;
+
+    @JsonProperty(value = FIELD_NAME_STATE_TTL, index = 1)
+    private final long stateTtl;
+
+    @JsonProperty(value = FIELD_NAME_STATE_NAME, index = 2)
+    private final String stateName;
+
+    @JsonCreator
+    public StateMetadata(
+            @JsonProperty(FIELD_NAME_STATE_INDEX) int stateIndex,
+            @JsonProperty(FIELD_NAME_STATE_TTL) long stateTtl,
+            @JsonProperty(FIELD_NAME_STATE_NAME) String stateName) {
```

Comment:
please check the `stateName ` is not null, otherwise `tateName.equals(that.stateName);` in `equals` method will cause NPE